### A. Introduction
####    A.1. Description & Disscusion of the Background


<br>
<p> Surulere is a residential and commercial Local Government Area located on the mainland of Lagos in Lagos State, Nigeria, with an area of 23 km². At the last census in the year 2006, there were 503,975 inhabitants, with a population density of 21,864 inhabitants per square kilometer. The local government area is bordered by Yaba, Mushin and Ebute-Metta.<sup>1</sup> Surulere is often touted as being the centre of lagos .</p>

What makes surulere lively is due to it's incongrous mixture in many ramifications. It boasts a  a fine blend between commercial and residential activities, a motley of upscale neighborhoods as well as midrange and even lowscale localities.it is home to a  mix of upscale cocktail bars, pulsing clubs, and laid-back beer parlors.

Like every city in Nigeria, Surulere is also plagued with epileptic power supply affecting small scale and big businesses alike , and is a major concern for both residential and commercial stakeholders in the city. Electricity shortages often mean residents resort to their petrol/diesel powered generator having to put up with the high blaring noise and it's detrimental effect on the environment especially at night when it is common to have as much as 100 generator within an 100m radius. companies spend upto  a shocking 70% of the budget on diesel or petrol<sup>2</sup>. 

There seems to be some hope for residents with the proposed partnership of renewable energy companies with  power distribution companies to provide alternative energy for consumers during power outage hours via an interconnected minigrid these  mini-grid will leverage on existing distribution infrastructure to achieve lower system cost than isolated mini-grid while improving service reliability from the status quo.

A mini grid, also sometimes referred to as a "micro grid or isolated grid", can be defined as a set of electricity generators and possibly energy storage systems interconnected to a distribution network that supplies electricity to a localized group of customers."They involve small-scale electricity generation (10 kW to 10MW) which serves a limited number of consumers via a distribution grid that can operate in isolation from national electricity transmission networks."<sup>3</sup>

“These systems can help DisCos reduce losses to serve these communities (by N1–2 billion per DisCo per year) and create a mini-grid market with N400 billion in annual revenue, all while saving customers ₦60 billion per year compared to what they spend today,”<sup>4</sup>. Mini Grids provide an enhanced service where there are epileptic power supply in urban areas too especially within the healthcare, higher education, government, businesses and the military base.

The aim of this project is to develop a clustering  model for the distribution of  mini-grids around surulere .Based on the nature of minigrids to serve only a limited number of customers , it is important for stakeholders to have an idea of the size requirements of each mini-grids that should be deployed to each locale . Considering the incongrous mixture of commercial,industrial and residential venues within surulere we can create a map that categorises the needs of each locale into a few clusters as well as an information chart about the types of business found within each locale  for easy visualization and better decision making.



####  A.2. Data Description
<br>

To consider the problem we can list the datas as below:

* The  Open Street map (OSM) API was used to find the co-ordinates and street names for all streets in Surulere, this would be important to give a final breakdown of the streets that fall into each grid .
* Note : The OSM API failed to give street names for some locations given only latitude and longitude and the reverse being the case for some location . we therefore rely on the Google API(find) to get this street names for locations without street names

   Data of the needed Columns  from the OSM API is returned as seen below .

* The location data obtained was cleaned carefully to only include data from surulere .

* After splitting our data into contiguos 300metres grids . We use the Google API to obtain the  location of our newly defined grid centers by passing the latitude and longitude of our new grids into the goole Places API (Reverse Geocoding)

* The google maps Places API(nearbysearch) was used to get the name and business type of every contiguos grid/locale within surulere. Our grids are defined as circular areas with a radius of 300 meters, so our grid centers will be 600 meters apart., google places API was also used to get the location and names of streets that we were not able to identify via the OSM API.

* Below is a sample of data obtained from each grid.


### B. Data 

Let's create latitude & longitude coordinates for centroids of our candidate neighborhoods. We will create a grid of cells covering our area of interest
Let's first find the latitude & longitude of our approximate center of surulere using specific, well known address and Google Maps geocoding API

In [2]:
import requests

def get_coordinates(api_key, address, verbose=False):
    try:
        url = 'https://maps.googleapis.com/maps/api/geocode/json?key={}&address={}'.format(api_key, address)
        response = requests.get(url).json()
        if verbose:
            print('Google Maps API JSON result =>', response)
        results = response['results']
        geographical_data = results[0]['geometry']['location'] # get geographical coordinates
        lat = geographical_data['lat']
        lon = geographical_data['lng']
        return [lat, lon]
    except:
        return [None, None]
    
address = 'Fountain height school, surulere, Lagos'
surulere_center = get_coordinates('xxxxxxxx,address)
print('Coordinate of {}: {}'.format(address, surulere_center))

Coordinate of Fountain height school, surulere, Lagos: [6.4936877, 3.3487219]



<iframe width="425" height="350" frameborder="0" scrolling="no" marginheight="0" marginwidth="0" src="https://www.openstreetmap.org/export/embed.html?bbox=3.2963275909423833%2C6.453226032208117%2C3.4096240997314458%2C6.524264986696486&amp;layer=mapnik" style="border: 1px solid black"></iframe><br/><small><a href="https://www.openstreetmap.org/#map=14/6.4887/3.3530">View Larger Map</a></small>

Let's Import list of all streets in surulere downloaded from open Street Map .

In [4]:
import json
import pandas as pd
data = json.load(open(r'C:\Users\SteelSeries\Desktop\export.json'))

In [10]:
from pandas import json_normalize

comp_view = json_normalize(data['elements'])
street_view = comp_view[['type','tags.name','lat','lon']]
street_way=street_view[street_view['type']=='way']
street_loc=street_view[street_view['type']=='node']

In [11]:
comp_view

type          id                                              nodes  \
0      way     5150272  [31674845, 5653922204, 5653922205, 2400903267,...   
1      way     8064810                             [84627688, 2400903208]   
2      way     9719406  [2251378632, 5187309417, 5187309420, 518730941...   
3      way    10219985  [5187309384, 91273418, 5624707091, 91273411, 5...   
4      way    10219988  [1802512421, 5624600330, 1802512423, 559027544...   
...    ...         ...                                                ...   
5976  node  5590376786                                                NaN   
5977  node  5594857669                                                NaN   
5978  node  5594857670                                                NaN   
5979  node  5594857671                                                NaN   
5980  node  5594857672                                                NaN   

     tags.bridge tags.highway tags.lanes tags.layer  \
0            yes     motorway          2          1   
1            NaN     motorway          2        NaN   
2            NaN     tertiary          2        NaN   
3            NaN  residential        NaN        NaN   
4            NaN    secondary        NaN        NaN   
...          ...          ...        ...        ...   
5976         NaN          NaN        NaN        NaN   
5977         NaN          NaN        NaN        NaN   
5978         NaN          NaN        NaN        NaN   
5979         NaN          NaN        NaN        NaN   
5980         NaN          NaN        NaN        NaN   

                      tags.name tags.oneway tags.ref tags.surface  \
0                Western Avenue         yes       A1        paved   
1                Western Avenue         yes       A1        paved   
2     Adeniran Ogunsanya Street         yes      NaN          NaN   
3        Modupe Johnson Cresent         NaN      NaN          NaN   
4        Babs Animashaun Street         NaN      NaN          NaN   
...                         ...         ...      ...          ...   
5976                        NaN         NaN      NaN          NaN   
5977                        NaN         NaN      NaN          NaN   
5978                        NaN         NaN      NaN          NaN   
5979                        NaN         NaN      NaN          NaN   
5980                        NaN         NaN      NaN          NaN   

                  tags.is_in tags.created_by tags.source tags.access  \
0                        NaN             NaN         NaN         NaN   
1                        NaN             NaN         NaN         NaN   
2     Surulere,Lagos,Nigeria             NaN         NaN         NaN   
3     Surulere,Lagos,Nigeria             NaN         NaN         NaN   
4     Surulere,Lagos,Nigeria             NaN         NaN         NaN   
...                      ...             ...         ...         ...   
5976                     NaN             NaN         NaN         NaN   
5977                     NaN             NaN         NaN         NaN   
5978                     NaN             NaN         NaN         NaN   
5979                     NaN             NaN         NaN         NaN   
5980                     NaN             NaN         NaN         NaN   

     tags.junction       lat       lon  
0              NaN       NaN       NaN  
1              NaN       NaN       NaN  
2              NaN       NaN       NaN  
3              NaN       NaN       NaN  
4              NaN       NaN       NaN  
...            ...       ...       ...  
5976           NaN  6.515141  3.369239  
5977           NaN  6.511785  3.370531  
5978           NaN  6.513121  3.370008  
5979           NaN  6.513675  3.369799  
5980           NaN  6.514128  3.369648  

[5981 rows x 18 columns]

Now let's create a grid of area candidates, equaly spaced, centered around city centre(Fountain  Heights School, surulere), and within ~3.5km from this center . Our neighborhoods will be defined as contiguos circular areas with a radius of 300 meters, so our neighborhood centers will be 600 meters apart. To accurately calculate distances we need to create our grid of locations in Cartesian 2D coordinate system which allows us to calculate distances in meters (not in latitude/longitude degrees). Then we'll project those coordinates back to latitude/longitude degrees to be shown on Folium map. So let's create functions to convert between WGS84 spherical coordinate system (latitude/longitude degrees) and UTM Cartesian coordinate system (X/Y coordinates in meters).

In [12]:
#!pip install shapely
import shapely.geometry

#!pip install pyproj
import pyproj

import math

def lonlat_to_xy(lon, lat):
    proj_latlon = pyproj.Proj(proj='latlong',datum='WGS84')
    proj_xy = pyproj.Proj(proj="utm", zone=33, datum='WGS84')
    xy = pyproj.transform(proj_latlon, proj_xy, lon, lat)
    return xy[0], xy[1]

def xy_to_lonlat(x, y):
    proj_latlon = pyproj.Proj(proj='latlong',datum='WGS84')
    proj_xy = pyproj.Proj(proj="utm", zone=33, datum='WGS84')
    lonlat = pyproj.transform(proj_xy, proj_latlon, x, y)
    return lonlat[0], lonlat[1]

def calc_xy_distance(x1, y1, x2, y2):
    dx = x2 - x1
    dy = y2 - y1
    return math.sqrt(dx*dx + dy*dy)

print('Coordinate transformation check')
print('-------------------------------')
print('surulere center longitude={}, latitude={}'.format(surulere_center[1], surulere_center[0]))
x, y = lonlat_to_xy(surulere_center[1], surulere_center[0])
print('surulere center UTM X={}, Y={}'.format(x, y))
lo, la = xy_to_lonlat(x, y)
print('surulere center longitude={}, latitude={}'.format(lo, la))

Coordinate transformation check
-------------------------------
surulere center longitude=3.3487219, latitude=6.4936877
surulere center UTM X=-797030.4725273035, Y=732850.4270195634
surulere center longitude=3.3487218999999966, latitude=6.493687699999999


Let's create a hexagonal grid of cells: we offset every other row, and adjust vertical row spacing so that every cell center is equally distant from all it's neighbors.

In [13]:
surulere_center_x, surulere_center_y = lonlat_to_xy(surulere_center[1], surulere_center[0]) # City center in Cartesian coordinates

k = math.sqrt(3) / 2 # Vertical offset for hexagonal grid cells
x_min = surulere_center_x - 3500
x_step = 600
y_min = surulere_center_y - 3500 - (int(21/k)*k*600 - 7000)/2
y_step = 600 * k 

latitudes = []
longitudes = []
distances_from_center = []
xs = []
ys = []
for i in range(0, int(21/k)):
    y = y_min + i * y_step
    x_offset = 300 if i%2==0 else 0
    for j in range(0, 21):
        x = x_min + j * x_step + x_offset
        distance_from_center = calc_xy_distance(surulere_center_x, surulere_center_y, x, y)
        if (distance_from_center <= 3351):
            lon, lat = xy_to_lonlat(x, y)
             
            latitudes.append(lat)
            longitudes.append(lon)
            distances_from_center.append(distance_from_center)
            xs.append(x)
            ys.append(y)

print(len(latitudes), 'candidate neighborhood centers generated.')

113 candidate neighborhood centers generated.


Let's visualize the data we have so far: city center location and candidate neighborhood centers in continguos grid around surulere.

In [15]:
import folium
map_surulere = folium.Map(location=surulere_center, zoom_start=13)
folium.Marker(surulere_center, popup='Surulere').add_to(map_surulere)
for lat, lon in zip(latitudes, longitudes):
    #folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_surulere) 
    folium.Circle([lat, lon], radius=300, color='blue', fill=False).add_to(map_surulere)
    #folium.Marker([lat, lon]).add_to(map_surulere)
map_surulere

We observe that Our grids overlap several locations outside Surulere. Therefore it is important that we manually filter out some grids .

In [16]:
import numpy as np
grid = []
for items in zip(latitudes, longitudes) :
    grid.append(items)
indices = np.array([ 97, 100, 101, 102, 103, 104, 105, 106, 107, 108,  14,  15,  16,
        17,  18,  19,  20,  83,  84,  93,  94,  95,  96,  62,  51,  73,
        40,  30,  11,   0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
        10, 110, 111, 112, 109])

surulere_grid = [v for i, v in enumerate(grid) if i not in indices]

In [17]:
map_surulere = folium.Map(location=surulere_center, zoom_start=13)
folium.Marker(surulere_center, popup='Surulere').add_to(map_surulere)
for lat, lon in surulere_grid:
    #folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_surulere) 
    folium.Circle([lat, lon], radius=300, color='blue', fill=False).add_to(map_surulere)
    #folium.Marker([lat, lon]).add_to(map_surulere)
map_surulere

Now our new grids show a more accurate representation of Surulere.
OK, we now have the coordinates of centers of neighborhoods/areas to be evaluated, equally spaced (distance from every point to it's neighbors is exactly the same) carefully fitted within surulere

Let's now use Google Maps API to get approximate addresses of those locations.

In [18]:
api_key = 'xxxxxxxxxxxxx'
def get_address(api_key, latitude, longitude, verbose=False):
    try:
        url = 'https://maps.googleapis.com/maps/api/geocode/json?key={}&latlng={},{}'.format(api_key, latitude, longitude)
        response = requests.get(url).json()
        if verbose:
            print('Google Maps API JSON result =>', response)
        results = response['results']
        address = results[0]['formatted_address']
        return address
    except:
        return None



In [20]:
print('Obtaining location addresses: ', end='')
addresses = []
for lat, lon in zip(latitudes, longitudes):
    address = get_address(api_key, lat, lon)
    if address is None:
        address = 'NO ADDRESS'
    address = address.replace(', Lagos', '') # We don't need country part of address
    addresses.append(address)
    print(' .', end='')
print(' done.')

Obtaining location addresses:  . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . done.


In [21]:
addresses_ref = [v for i, v in enumerate(addresses) if i not in indices]
latitudes_ref = [v for i, v in enumerate(latitudes) if i not in indices]
longitudes_ref = [v for i, v in enumerate(longitudes) if i not in indices]
xs_ref = [v for i, v in enumerate(xs) if i not in indices]
ys_ref = [v for i, v in enumerate(ys) if i not in indices]
distances_from_center_ref = [v for i, v in enumerate(distances_from_center) if i not in indices]

Now , we have successfully limited our data to surulere only and obtained addresses for each of our grid centres

In [23]:
addresses[1:10]

['18 Oyegbemi St, Malu, Nigeria',
 '132 Igbobi College Rd, Badia, Nigeria',
 '73 Gasikia Rd, Badia, Nigeria',
 '30 Igbaja St, Amukoko, Nigeria',
 '20 Oguniyi St, Amukoko, Nigeria',
 '8 Mission St, Amukoko, Nigeria',
 '17 Iganmu Rd, Amukoko, Nigeria',
 '38 Opeloyeru St, Ijora, Nigeria',
 '30 Bale St, Ijora, Nigeria']

So , Let's now place all this into a Pandas dataframe.

In [24]:
import pandas as pd

df_locations = pd.DataFrame({ 'Addresses': addresses_ref,
                             'Latitude': latitudes_ref,
                             'Longitude': longitudes_ref,
                             'X': xs_ref,
                             'Y': ys_ref,
                             'Distance from center': distances_from_center_ref})

df_locations[['Addresses','Latitude','Longitude']].head(10)

Addresses  Latitude  Longitude
0                        95 Coker Rd, Coker, Nigeria  6.474863   3.331438
1                          2 Ecwa Rd, Coker, Nigeria  6.474987   3.336752
2                      11 Abo-Aba Cl, Coker, Nigeria  6.479404   3.328674
3                     26 Coker Rd, Surulere, Nigeria  6.479528   3.333988
4                 88 Baale St, Orile Iganmu, Nigeria  6.479652   3.339302
5                19 Savage St, Orile Iganmu, Nigeria  6.479776   3.344615
6               31 Jimoh Odutola St, Iganmu, Nigeria  6.479899   3.349929
7  KM2 Lagos - Badagry Expy, Surulere 100242, Nig...  6.480023   3.355243
8                64 Eric Moore Rd, Abulenla, Nigeria  6.480147   3.360557
9                         Brewery, Abulenla, Nigeria  6.480270   3.365872

Next , We use the Google Maps Places API to obtain data on nearby venues , We are interested in the names of this location and type for venues that are businesses . 
The google API refers to business in general as establishments and gives a more detailed (type) for few locations .

In [25]:

results = []
for lat,lon in zip(df_locations['Latitude'],df_locations['Longitude']) :
    url = 'https://maps.googleapis.com/maps/api/place/nearbysearch/json?location={},{}&radius=450&key=xxxxxxxxxx'.format(lat,lon)
    response = requests.get(url).json()
    result = response['results']
    results.append(result)
    

We now have the nearby venues each of our grid centers. 
The next step is to extract the needed information from this data

In [26]:
def get_nearby_venues(address,latitude,longitude,x,y,distance):
    
    results = []
    for address,latitude,longitude,x,y,distance in zip(address,latitude,longitude,x,y,distance) :
        url = 'https://maps.googleapis.com/maps/api/place/nearbysearch/json?location={},{}&radius=450&key=xxxxxxxxxxxxxxxxxxxxxx'.format(latitude,longitude)
        response = requests.get(url).json()
        result = response['results']
        results.append([(address,
                      latitude,
                      longitude,
                      x,
                      y,
                      distance,
                         item['geometry']['location']['lat'],
                         item['geometry']['location']['lng'],                         
                      item['name'],
                      item['types']) for item in result])
    nearby_venues = pd.DataFrame([items for venues in results for items in venues])
    nearby_venues.columns = ['Addresses',
                            'Latitude',
                            'Longitude',
                            'X',
                            'Y',
                            'Distance_from_center',
                             'ven_lat',
                             'ven_lon',
                            'Venue',
                            'Type']
    return(nearby_venues)

In [27]:
surulere_venues = get_nearby_venues(address= df_locations['Addresses'],
                  latitude=df_locations['Latitude'],
                  longitude= df_locations['Longitude'],
                  x= df_locations['X'],
                  y=df_locations['Y'],
                  distance=df_locations['Distance from center'])

In [61]:
 surulere_venues = surulere_venues_deep.copy()

In [62]:
#We create a copy of our data to avoid calling over the google api severally.in the event of an incidence with the data
surulere_venues_deep = surulere_venues.copy()

Let's clean our data for easier understanding .

In [63]:
import re
surulere_venues['Type'] = [','.join(map(str, l)) for l in surulere_venues['Type']]
surulere_venues['Type'] = [items.replace('point_of_interest,','') for items in surulere_venues['Type']]
surulere_venues['Type'] = [items.replace(',establishment','') for items in surulere_venues['Type']]
surulere_venues['Type'] = [re.sub(',.*$','',items) for items in surulere_venues['Type']]
surulere_venues[['Venue','Type']];

Let's now see all the collected venues in our area of interest on map within their respective clusters.

In [65]:
suru_map = folium.Map(surulere_center, zoom_start=13)
folium.Marker(surulere_center,popup=folium.Popup('Surulere,Lagos')).add_to(suru_map)

for ven in surulere_venues.values:
    lat = ven[-4]
    lon = ven[-3]
    label = ven[-1]
    
    folium.CircleMarker([lat,lon], radius = 0.5,
                       color = 'red', popup = label).add_to(map_surulere)

map_surulere

the map above gives us a clue of what our analysis will be predicated upon.We hope to be able to reach a conclusion on the most suitable grid location for our mini grid solutions . based on the venues within each grid as well as venues just outside the grid for a more accurate result.

In [66]:
surulere_venues.describe(include='all')

Addresses     Latitude    Longitude  \
count                                  1380  1380.000000  1380.000000   
unique                                   69          NaN          NaN   
top     17 Adeyemi Rd, Ebute Metta, Nigeria          NaN          NaN   
freq                                     20          NaN          NaN   
mean                                    NaN     6.493542     3.351076   
std                                     NaN     0.009786     0.014240   
min                                     NaN     6.474863     3.325695   
25%                                     NaN     6.484812     3.338872   
50%                                     NaN     6.493646     3.351835   
75%                                     NaN     6.502604     3.362893   
max                                     NaN     6.512307     3.376286   

                    X              Y  Distance_from_center      ven_lat  \
count     1380.000000    1380.000000           1380.000000  1380.000000   
unique            NaN            NaN                   NaN          NaN   
top               NaN            NaN                   NaN          NaN   
freq              NaN            NaN                   NaN          NaN   
mean   -796765.255136  732827.835053           1830.237826     6.495356   
std       1610.500901    1099.771106            722.434802     0.011561   
min    -799630.472527  730771.966050            200.000000     6.470891   
25%    -798130.472527  731811.196535           1311.487705     6.486029   
50%    -796630.472527  732850.427020           1907.878403     6.494892   
75%    -795430.472527  733889.657504           2357.965225     6.502644   
max    -793930.472527  734928.887989           3143.246729     6.524379   

            ven_lon  Venue           Type  
count   1380.000000   1380           1380  
unique          NaN   1029             60  
top             NaN  Lagos  establishment  
freq            NaN     69            648  
mean       3.352519    NaN            NaN  
std        0.015009    NaN            NaN  
min        3.322930    NaN            NaN  
25%        3.339970    NaN            NaN  
50%        3.351428    NaN            NaN  
75%        3.364112    NaN            NaN  
max        3.380246    NaN            NaN

So far, We have been able to gather data of 1380 venues , from 69 grids within Surulere,  Next is to prepare our data for Processing . 